### PR TITLE
Configurable api client

### PR DESCRIPTION
**Pull request by shylesh**
This PR has 2 commits. 
1. Which handles the api-client configurations at run time
2. A simple test for api-client infrastructure

### CODE REVIEW

File: ocs/api_client.py [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+    This implements all functionalities like create, patch, delete using
+    oc commands.
+    """
+
+    @property
+    def name(self):
+        return "OCCLIClient"
+
+
+class KubClient(APIClientBase):
```
Comment (vasukulkarni) [inline ocs/api_client.py]: why not KubeClient?

File: ocs/api_client.py [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+    """
+    All activities using upstream kubernetes python client
+    """
+
+    @property
+    def name(self):
+        return "KubClient"
```
Comment (vasukulkarni) [inline ocs/api_client.py]: s/KubClient/KubeClient

**Comment by vasukulkarni** [comment]
Other than minor nits, LGTM,  Thanks for abstracting out functions into Kube vs Openshift which will make it easy to run on both environments.

File: ocs/api_client.py (hunk 2 above)
Comment (shylesh) [reply ocs/api_client.py]: Done

File: ocs/api_client.py (hunk 1 above)
Comment (shylesh) [reply ocs/api_client.py]: Done

File: ocs/api_client.py [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+
+            Note: Returns could be tricky because it varies from client to
+            client. If its oc-cli then extra work needs to be done in the
+            specific implementation.
+        """
+
+        raise NotImplementedError("get_labels method is not implemented")
+
+
+# All openshift REST client specific imports here
```
Comment (petr-balogh) [inline ocs/api_client.py]: I would move all imports at top of module and order them in pep8 way.

File: ocs/api_client.py [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+
+        resource = self.rest_client.v1_services
+        kw.update({'resource': resource})
+        return self.api_create(**kw)
+
+    def api_get(self, **kw):
+        return self.rest_client.call_api("GET", **kw)
+
+    def api_post(self, **kw):
+        pass
```
Comment (petr-balogh) [inline ocs/api_client.py]: super(). api_post(**kw) - this way we will raise NotImplementedError as you have defined in abstract method.  The same below.

File: ocs/api_client.py [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+class OCRESTClient(APIClientBase):
+    """ All activities using openshift REST client"""
+
+    def __init__(self):
+        """TODO: get REST client instance from ctx which is shared globally"""
+        self.rest_client = openshift_ops.OCP()
+
+    @property
+    def name(self):
+        return "OCRESTClient"
```
Comment (petr-balogh) [inline ocs/api_client.py]: what about 
```python
return self.__class__.__name__
```

**Comment by petr-balogh** [comment]
And please once WIP is over, remove it from the name. Thanks

File: oc/openshift_ops.py [hunk 6]
```diff
@@ -25,6 +25,38 @@ def __init__(self):
         self.v1_pods = dyn_client.resources.get(
             api_version='v1', kind='Pod'
         )
+        self.v1_deployments = dyn_client.resources.get(
+            api_version='v1', kind='Deployment'
+        )
+        self.v1_services = dyn_client.resources.get(
```
Comment (RazTamir) [inline oc/openshift_ops.py]: Can we remove the v1 from the variable name?
The api version is v1 but the name shouldn't mention that

+ can we maintain the 'v1' in the defaults so as soon as the api version with change, we will be able to change only 1 parameter?

**Comment by shylesh** [comment]
> Does `test_api_client.py` require a running cluster to work properly or would this be better suited as a unit test? Right now it's included in the same module as our product tests and I don't think it belongs there as it's testing part of our framework.

This test assumes that openshift cluster is up and running and creates an example service although it has nothing specific to rook cluster. However this is exercising the api-client interface which we intend to use for our general activities which involves oc create, patch, post etc. 

I am not sure whether we can keep it here or move it to some other place, do you feel it should be kept seperate ?

**Comment by clacroix12** [comment]
> I am not sure whether we can keep it here or move it to some other place, do you feel it should be kept seperate ?

I'm more just trying to understand the nature of this test, and when we plan to execute it. It doesn't look like it's been added to any of our suites and seems like it would be better as a unit test if anything.

File: ocs/api_client.py (hunk 3 above)
Comment (shylesh) [reply ocs/api_client.py]: Done

File: ocs/api_client.py (hunk 4 above)
Comment (shylesh) [reply ocs/api_client.py]: Done

File: ocs/api_client.py (hunk 5 above)
Comment (shylesh) [reply ocs/api_client.py]: Actually I was not sure whether config will have same parameter value as class name hence kept that string. However with this change user has to use same value as class name.

Done

File: oc/openshift_ops.py (hunk 6 above)
Comment (shylesh) [reply oc/openshift_ops.py]: Done.

**Comment by shylesh** [comment]
Looks like I have merge conflicts. Once if you guys agree on review related changes I will rebase and resolve the conflict.

**Comment by petr-balogh** [comment]
@shylesh go ahead with rebase and solve the conflict.

**Comment by vasukulkarni** [comment]
@shylesh please rebase

**Comment by clacroix12** [comment]
LGTM

**Pull request merged**